contracts/0.8.9/oracle/ValidatorsExitBusOracle.sol

```diff
@@ -0,0 +1,507 @@
+        uint256 /* prevSubmittedRefSlot */,
+        uint256 prevProcessingRefSlot
+    ) internal override {
+        DataProcessingState memory state = _storageDataProcessingState().value;
+        if (state.refSlot == prevProcessingRefSlot && state.requestsProcessed < state.requestsCount) {
+            emit WarnDataIncompleteProcessing(
+                prevProcessingRefSlot,
+                state.requestsProcessed,
+                state.requestsCount
+            );
+        }
+    }
+
+    function _checkMsgSenderIsAllowedToSubmitData() internal view {
+        address sender = _msgSender();
+        if (!hasRole(SUBMIT_DATA_ROLE, sender) && !_isConsensusMember(sender)) {
+            revert SenderNotAllowed();
+        }
+    }
+
+    function _handleConsensusReportData(ReportData calldata data) internal {
+        if (data.dataFormat != DATA_FORMAT_LIST) {
+            revert UnsupportedRequestsDataFormat(data.dataFormat);
+        }
+
+        if (data.data.length % 64 != 0) {
+            revert InvalidRequestsDataLength();
+        }
+
+        if (data.data.length / 64 != data.requestsCount) {
+            revert UnexpectedRequestsDataLength();
+        }
+
+        uint256 lastProcessedItemWithoutPubkey = _processExitRequestsList(data.data);
+
+        _storageDataProcessingState().value = DataProcessingState({
+            lastProcessedItemWithoutPubkey: lastProcessedItemWithoutPubkey,
+            refSlot: data.refSlot.toUint64(),
+            requestsCount: data.requestsCount.toUint64(),
+            requestsProcessed: data.requestsCount.toUint64(),
+            dataFormat: uint16(DATA_FORMAT_LIST)
+        });
+
+        if (data.requestsCount == 0) {
+            return;
+        }
+
+        RateLimit
+            .load(RATE_LIMIT_POSITION)
+            .recordUsageAt(data.refSlot, data.requestsCount * 10**18)
+            .store(RATE_LIMIT_POSITION);
+
+        TOTAL_REQUESTS_PROCESSED_POSITION.setStorageUint256(
+            TOTAL_REQUESTS_PROCESSED_POSITION.getStorageUint256() + data.requestsCount
+        );
+    }
+
+    function _processExitRequestsList(bytes calldata data) internal returns (uint256) {
+        uint256 offset;
+        uint256 offsetPastEnd;
+        assembly {
+            offset := data.offset
+            offsetPastEnd := add(offset, data.length)
+        }
+
+        mapping(uint256 => RequestedValidator) storage _lastReqValidatorIndices =
+            _storageLastRequestedValidatorIndices();
+
+        uint256 lastDataWithoutPubkey = 0;
+        uint256 lastNodeOpKey = 0;
+        uint256 lastValIndex;
+        bytes calldata pubkey;
+
+        assembly {
+            pubkey.length := 48
+        }
+
+        while (offset < offsetPastEnd) {
+            uint256 dataWithoutPubkey;
+            assembly {
+                // 16 most significant bytes are taken by module id, node op id, and val index
+                dataWithoutPubkey := shr(128, calldataload(offset))
```

Comment:
why not using slices?

Reply:
like here https://github.com/lidofinance/lido-dao/blob/master/contracts/0.6.11/deposit_contract.sol#L131

Reply:
good catch, forgot that Solidity has them now. will rewrite using slices

Reply:
any updates?